### PR TITLE
fix: correctly handle ISR for appDir pages

### LINKED ISSUE #1844: [Bug]: appDir pages not varying cache on RSC header
### Summary

I tried fetching in Postman and *sometimes* Netlify will make user download a file with following content: https://pastebin.com/raw/QMZ7wVuB

which I assume is Next.js's server component props

### Steps to reproduce

View the site

### A link to a reproduction repository

https://github.com/Enime-Project/enime.moe/tree/next

### Next Runtime version

4.29.3-appdir.0

### More information about your build

- [ ] I am building using the CLI
- [ ] I am building using file-based configuration (`netlify.toml`)

### What OS are you using?

None

### Your netlify.toml file

<details>
  <summary>`netlify.toml`</summary>

```toml
# Paste content of your `netlify.toml` file here
```

</details>

### Your public/_redirects file

<details>
  <summary>`_redirects`</summary>

```toml
# Paste content of your `_redirects` file here
```

</details>

### Your `next.config.js` file

<details>
  <summary>`next.config.js`</summary>

```toml
# Paste content of your `next.config.js` file here. Check there is no private info in there.
```

</details>

### Builds logs (or link to your logs)

<details>
  <summary>Build logs</summary>

```
# Paste logs here
```

</details>

### Function logs

<details>
  <summary>Function logs</summary>

```
# Paste logs here
```

</details>

### .next JSON files

<details>
  <summary>generated .next JSON files</summary>

```
# Paste file contents here. Please check there isn't any private info in them
# You can either build locally, or download the deploy from Netlify by clicking the arrow next to the deploy time.
```

</details>

### PR DESCRIPTION
**Pull request by ascorbic**
### Summary
Data routes are handled differently for appDir pages. Instead of a JSON file in `/_next/data`, the request is made to the same path, passing the `RSC` header. This returns the RSC data to stream the update. The fact this varies on a header means we can't use simple ODB handling for these pages, because it will lead to errors like #1844, where the RSC data is returned when HTML is requested, or vice versa. 

This PR fixes this by adding an edge function which routes appDir requests that include the `rsc` header to the appropriate `.rsc` path instead. The original request path is put into a header, which then replaces the `.rsc` path in the handler. These are real files but not normally requested directly. The reason for doing this is so that these requests have unique URLs, meaning they're cached separately as ODBs. 

There's a final extra part that is needed, which is to handle trailing slashes. If we normalise all of the requests to the matching `.rsc` file, then there will be no difference to requests with and without the trailing slash. This is a problem, because depending on whether the version with or without the slash is requested first, we may cache the redirect instead of the actual data (or vice versa). To handle this, we do a bit of trickery. If the request has a trailing slash, we include a trailing slash on the rewritten `.rsc`path. For example, while an rsc request to `/blog` will be rewritten to `/blog.rsc`, a request for `/blog/` (with trailing slash) will be rewritten to `/blog.rsc/`. This means that it will be treated as a separate ODB path, and will be cached differently.

Basically, all of this is to implement `Vary` in ODBs. 🤷🏻 

### Test plan

1. Visit [an appDir page on the default demo site](https://deploy-preview-1855--netlify-plugin-nextjs-demo.netlify.app/blog/erica/). Navigate around, checking in the network panel that SPA data requests are correctly returning `application/octet-stream` RSC data. 
2. Load any of the `/blog` pages directly and ensure that HTML is returned
3. Try loading them with and without the trailing slash and see if they redirect correctly.

### Relevant links (GitHub issues, Notion docs, etc.) or a picture of cute animal
Fixes #1844

capy wants an apple

### Standard checks:

- [ ] Check the Deploy Preview's Demo site for your PR's functionality
- [ ] Add docs when necessary

---

🧪 Once merged, make sure to update the version if needed and that it was published correctly.

### CODE REVIEW

**Comment by NADESHIKON** [comment]
When is this getting merged?

**Comment by NADESHIKON** [comment]
When is this getting merged?

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -15,32 +15,19 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu-latest, macOS-latest, windows-latest]
-        node-version: [14, '*']
```
Comment (ascorbic) [inline .github/workflows/test.yml:18]: AppDir requires 16+

File: test/__snapshots__/index.spec.js.snap [hunk 2]
```diff
@@ -54,9 +60,9 @@ exports[`onBuild() generates a file referencing all API route sources: for _api_
 exports.resolvePages = () => {
     try {
         require.resolve('../../../.next/package.json')
-        require.resolve('../../../.next/server/chunks/274.js')
-        require.resolve('../../../.next/server/chunks/7016.js')
-        require.resolve('../../../.next/server/chunks/7590.js')
+        require.resolve('../../../.next/server/chunks/CHUNK_ID.js')
```
Comment (ascorbic) [inline test/__snapshots__/index.spec.js.snap:63]: These chunk name snapshots are too flaky

File: test/__snapshots__/index.spec.js.snap [hunk 3]
```diff
@@ -1265,6 +1295,18 @@ Array [
     "status": 404,
     "to": "/404.html",
   },
+  Object {
+    "force": false,
+    "from": "/blog/:author",
+    "status": 200,
+    "to": "/.netlify/functions/___netlify-handler",
+  },
```
Comment (ascorbic) [inline test/__snapshots__/index.spec.js.snap:1303]: RSC pages use SSR

File: test/index.spec.js [hunk 4]
```diff
@@ -59,6 +59,8 @@ const utils = {
   },
 }
 
+const normalizeChunkNames = (source) => source.replaceAll(/\/chunks\/\d+\.js/g, '/chunks/CHUNK_ID.js')
```
Comment (ascorbic) [inline test/index.spec.js:62]: The chunk id is too easy to break between platforms, and shouldn't indicate a test failure. This normalises them all to "CHUNK_ID"

File: .github/workflows/e2e-appdir.yml [hunk 5]
```diff
@@ -41,7 +41,6 @@ jobs:
         env:
           NETLIFY_AUTH_TOKEN: ${{ secrets.NETLIFY_BOT_AUTH_TOKEN }}
           NETLIFY_SITE_ID: 1d5a5c76-d445-4ae5-b694-b0d3f2e2c395
-          NEXT_TEST_VERSION: canary
```
Comment (ascorbic) [inline .github/workflows/e2e-appdir.yml:44]: Run tests with the same version of next as the repo

File: cypress/integration/default/appdir.spec.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+  it('correctly redirects HTML requests for ISR pages', () => {
+    cy.request('/blog/erica').then((response) => {
+      expect(response.status).to.be('308')
+      expect(response.headers).to.have.property('location', '/blog/erica/')
+    })
+  })
+
+  // This needs trailing slash handling to be fixed
+  it.skip('correctly redirects HTML requests for static pages', () => {
```
Comment (ascorbic) [inline cypress/integration/default/appdir.spec.ts:61]: Prerendered pages do the normal Netlify trailing slash handling. This shoudl work once we have the trialing slash edge funciton

File: packages/runtime/src/helpers/redirects.ts [hunk 7]
```diff
@@ -142,30 +146,41 @@ const generateStaticIsrRewrites = ({
...
       staticIsrRewrites.push(
         // No i18n, because the route is already localized
-        ...redirectsForNextRoute({ route, basePath, to: ODB_FUNCTION_PATH, force: true, buildId, i18n: null }),
+        ...redirectsForNextRoute({
+          route,
+          basePath,
+          to: ODB_FUNCTION_PATH,
+          force: true,
+          buildId,
+          dataRoute: isAppDir ? dataRoute : null,
```
Comment (ascorbic) [inline packages/runtime/src/helpers/redirects.ts:181]: appDir routes provide their own data route value

**Comment by orinokai** [comment]
@ascorbic is there a reason we can't just not normalise the request? so a request to /blog will be rewritten to /blog.rsc and a request to /blog/ will be rewritten to /blog.rsc/

it would make it easier to reason about when looking at logs, etc. but let me know if i've misunderstood something

**Comment by ascorbic** [comment]
I tried that at first, and it didn't work. I can't for the life of me rememebr why! Let me see if I can reproduce it. You're right it would be simpler. I'm now wondering if other changes I made sinc ethen will make it work.

**Comment by ascorbic** [comment]
@orinokai OK, so it didn't work right away, but with a few changes it does! Good suggestion. I've updated the description too.

File: packages/runtime/src/helpers/utils.ts [hunk 8]
```diff
@@ -63,21 +63,38 @@ export const toNetlifyRoute = (nextRoute: string): Array<string> => {
...
-]
+export const generateNetlifyRoutes = ({
+  route,
+  dataRoute,
+  withData = true,
+}: {
+  route: string
+  dataRoute: string
+  withData: boolean
+}) => [...(withData ? toNetlifyRoute(dataRoute) : []), ...toNetlifyRoute(route)]
```
Comment (orinokai) [inline packages/runtime/src/helpers/utils.ts:74]: Will be so good to replace all these helper functions with edge router!

**Review by orinokai — APPROVED** [review]
Amazing! Nice one for sticking with it. 🚢

**Comment by gpoole** [comment]
@ascorbic the issue described in #1844 and fixed by this PR unfortunately seems to have surfaced again as #2089. It's happening with auto-detected version of the runtime with Next 13.5.6 (latest released v13 as of now). Could it be the same underlying problem and, if so, potentially a regression?

**Comment by ascorbic** [comment]
@gpoole yes, this is related to internal changes in Next.js. We're working on some major changes that will fix this and many more issues. Beta will be available soon.

**Comment by gpoole** [comment]
That's great, thanks for the update 👍
-------- Original Message --------

**Pull request merged**